osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs

```diff
@@ -1,19 +1,528 @@
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            drawableRuleset.Overlays.Add(new TargetBeatContainer());
+        }
+
+        public class TargetBeatContainer : BeatSyncedContainer
+        {
+            private PausableSkinnableSound sample;
+
+            public TargetBeatContainer()
+            {
+                Divisor = 1;
+            }
+
+            protected override void OnNewBeat(int beatIndex, TimingControlPoint timingPoint, EffectControlPoint effectPoint, ChannelAmplitudes amplitudes)
+            {
+                base.OnNewBeat(beatIndex, timingPoint, effectPoint, amplitudes);
+
+                if (!IsBeatSyncedWithTrack) return;
+
+                sample?.Play();
+            }
+
+            [BackgroundDependencyLoader]
+            private void load()
+            {
+                InternalChildren = new Drawable[]
+                {
+                    sample = new PausableSkinnableSound(new SampleInfo("Gameplay/nightcore-hat")) // todo: use another sample
+                };
+            }
+        }
+
+        #endregion
+
+        #region Helper Subroutines
+
+        /// <summary>
+        /// Get samples (if any) for a specific point in time.
+        /// </summary>
+        /// <remarks>
+        /// Samples will be returned if a hit circle or a slider node exists at that point of time.
+        /// </remarks>
+        /// <param name="hitObjects">The list of hit objects in a beatmap, ordered by StartTime</param>
+        /// <param name="time">The point in time to get samples for</param>
+        /// <returns>Hit samples</returns>
+        private IList<HitSampleInfo> getSamplesAtTime(IEnumerable<OsuHitObject> hitObjects, double time)
+        {
+            var sampleObj = hitObjects.FirstOrDefault(x =>
+            {
+                if (Precision.AlmostEquals(time, x.StartTime)) return true;
+
+                if (!(x is Slider s))
+                    return false;
+                if (!Precision.AlmostBigger(time, s.StartTime)
+                    || !Precision.AlmostBigger(s.EndTime, time))
+                    return false;
+
+                return Precision.AlmostEquals((time - s.StartTime) % s.SpanDuration, 0);
+            });
+            if (sampleObj == null) return null;
+
+            IList<HitSampleInfo> samples;
+
+            if (sampleObj is Slider slider)
+                samples = slider.NodeSamples[(int)Math.Round((time - slider.StartTime) % slider.SpanDuration)];
+            else
+                samples = sampleObj.Samples;
+
+            return samples;
+        }
+
+        private bool isOverlappingWithRecent(IReadOnlyList<OsuHitObject> hitObjects, int idx)
+        {
+            var target = hitObjects[idx];
+            return hitObjects.SkipLast(hitObjects.Count - idx).TakeLast(overlap_check_count)
+                             .Any(h => Vector2.Distance(h.Position, target.Position) < target.Radius * 2);
```

Comment:
Why is this operating on the entire list and iindirectly iterating over it via linq rather than taking a `Span` of the few objects that are actually needed here or something of the sort?

Reply:
I've refactored the method, but I'm not sure if you can take a `Span` from a `List`, so the method is accepting an `IEnumerable` for now.